File: 647.回文子串/solution.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        ans = 0
        dp = [[False] * n for _ in range(n)]
        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                if s[i] == s[j] and j - i <= 1:
                    dp[i][j] = True
                    ans += 1
                    continue
                if s[i] == s[j] and dp[i + 1][j - 1] == True:
                    dp[i][j] = True
                    ans += 1
        return ans

File: 647.回文子串/solution.cpp
#include "..\leetcode.h"

class Solution {
public:
    int countSubstrings(string s) {
        int n = s.size();
        int ans = 0;
        // dp[i][j] s[i,j]是不是回文子串
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        for(int i = n - 1; i >= 0; i--) {
            for(int j = i; j < n; j++) {
                // 一个数或者aa
                if(s[i] == s[j] && j - i <= 1) {
                    dp[i][j] = true;
                    ans++;
                    continue;
                }
                // i,j 依赖 i + 1,j - 1
                if(s[i] == s[j] && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                    ans++;
                }
            }
        }
        return ans;
    }
};